pkg/chains/factory.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2022 The Tekton Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package chains
+
+import (
+	"context"
+
+	"github.com/tektoncd/chains/pkg/chains/formats"
+	"github.com/tektoncd/chains/pkg/chains/formats/intotoite6"
+	"github.com/tektoncd/chains/pkg/chains/formats/provenance"
+	"github.com/tektoncd/chains/pkg/chains/formats/simple"
+	tekton_format "github.com/tektoncd/chains/pkg/chains/formats/tekton"
+	"github.com/tektoncd/chains/pkg/chains/signing"
+	"github.com/tektoncd/chains/pkg/chains/signing/kms"
+	"github.com/tektoncd/chains/pkg/chains/signing/x509"
+	"github.com/tektoncd/chains/pkg/chains/storage"
+	"github.com/tektoncd/chains/pkg/chains/storage/docdb"
+	"github.com/tektoncd/chains/pkg/chains/storage/gcs"
+	"github.com/tektoncd/chains/pkg/chains/storage/oci"
+	tekton_storage "github.com/tektoncd/chains/pkg/chains/storage/tekton"
+	"github.com/tektoncd/chains/pkg/config"
+	versioned "github.com/tektoncd/pipeline/pkg/client/clientset/versioned"
+	"go.uber.org/zap"
+	"k8s.io/client-go/kubernetes"
+)
+
+func AbstractFactory(ctx context.Context, pipelineClientset versioned.Interface, secretPath string, kc kubernetes.Interface, cfg config.Config, logger *zap.SugaredLogger) (map[formats.PayloadType]formats.Payloader, map[string]signing.Signer, map[string]storage.Backend, error) {
```

Comment:
what are the benefits of putting all of these in the AbstractFactory? why not call them separately?

Reply:
Just thinking it looks cleaner to have a single factory function to return all 3 objects we need to sign a taskrun. not sure if it's a common use case.
The other option is to keep `allFormatters` and `allSigners` as is in the `signing.go` as @wlynch suggested [here](https://github.com/tektoncd/chains/pull/398#discussion_r831569414). But those functions will be called in `controller.go` not in the `signing.go` file so we might need to export them. 

@wlynch what do you think about this?

Reply:
the only issue I can see is to couple these together, so if you only want to get backends somewhere else you would need to call the function and process all of them